l1-contracts/src/core/libraries/Hash.sol

```diff
@@ -31,28 +31,56 @@ library Hash {
     );
   }
 
+    /**
+   * @notice Computes the sha256 hash of the L1 to L2 message and converts it to a field element
+   * @param _message - The L1 to L2 message to hash
+   * @return The hash of the provided message as a field element
+   */
+  function sha256ToField32(DataStructures.L1ToL2Msg memory _message) internal pure returns (bytes32) {
+    return sha256ToField32(
+      abi.encode(
+        _message.sender,
+        _message.recipient,
+        _message.content,
+        _message.secretHash,
+        _message.deadline,
+        _message.fee
+      )
+    );
+  }
+
   /**
    * @notice Computes the sha256 hash of the L2 to L1 message and converts it to a field element
    * @param _message - The L2 to L1 message to hash
    * @return The hash of the provided message as a field element
    */
-  function sha256ToField(DataStructures.L2ToL1Msg memory _message) internal pure returns (bytes32) {
-    return sha256ToField(abi.encode(_message.sender, _message.recipient, _message.content));
+  function sha256ToField32(DataStructures.L2ToL1Msg memory _message) internal pure returns (bytes32) {
+    return sha256ToField32(abi.encode(_message.sender, _message.recipient, _message.content));
   }
 
   /**
    * @notice Computes the sha256 hash of the provided data and converts it to a field element
-   * @dev Using modulo to convert the hash to a field element.
+   * @dev Truncating one byte to convert the hash to a field element. We prepend a byte rather than cast bytes31(bytes32) to match Noir's to_be_bytes.
+   * @param _data - The bytes to hash
+   * @return The hash of the provided data as a field element
+   */
+  function sha256ToField32(bytes memory _data) internal pure returns (bytes32) {
+    return bytes32(bytes.concat(new bytes(1), bytes31(sha256(_data))));
```

Comment:
In integrating the truncated SHA into this PR, I discovered a discrepancy in byte casting between Solidity and Noir/ts.
e.g. say normal sha256 gives a result of `0x2e7ff14389eef3dc51597529149e01b49cb33829f7089438c8c145c8f352c17b`.
In Solidity the cleanest way it seems to drop a byte is `bytes31(thing)`. Turning it back into a `bytes32` for checks and structs gives:
`0x2e7ff14389eef3dc51597529149e01b49cb33829f7089438c8c145c8f352c100`.

However, both `toBuffer()` in ts and `to_be_bytes()` in Noir would give:
`0x002e7ff14389eef3dc51597529149e01b49cb33829f7089438c8c145c8f352c1`

The solidity version fails in-field checks since it still fills the most significant bytes, so I went for the Noir/TS version wherever checks are needed. That's why in some places `sha256ToField32` is used, and in others `sha256ToField` (which returns `bytes31`).
I also changed logs hashes to be treated as `bytes31` wherever possible to avoid any issues with casting. The above fixes are a bit janky so I'm happy for any feedback on them.